velox/exec/SortBuffer.cpp

```diff
@@ -0,0 +1,348 @@
+    spillSources_.resize(outputBatchSizeLimit_);
+    spillSourceRows_.resize(outputBatchSizeLimit_);
+
+    return std::optional(spillStats);
+  }
+}
+
+RowVectorPtr SortBuffer::output() {
+  VELOX_CHECK(noMoreInput_);
+  if (numRowsReturned_ == addedNumRows_) {
+    return nullptr;
+  }
+
+  // Return data in batches
+  initOrResetOutputBuffer();
+  if (sortedSpiller_ == nullptr) {
+    prepareOutput();
+  } else {
+    prepareOutputWithSpill();
+  }
+  return output_;
+}
+
+void SortBuffer::reclaim(uint64_t targetBytes) {
+  // TODO: support fine-grain disk spilling based on 'targetBytes' after having
+  // row container memory compaction support later.
+  spill(0, targetBytes);
+  VELOX_CHECK_EQ(data_->numRows(), 0);
+  data_->clear();
+  // Release the minimum reserved memory.
+  memoryPool_->release();
+}
+
+void SortBuffer::triggerSpillIfNeeded(const RowVectorPtr& input) {
+  // Check if spilling is enabled or not.
+  if (!spillConfig_.has_value()) {
+    return;
+  }
+
+  const int64_t currentDataRows = data_->numRows();
+  if (currentDataRows == 0) {
+    // 'data_' is empty. Nothing to spill.
+    return;
+  }
+
+  auto [freeRows, outOfLineFreeBytes] = data_->freeSpace();
+  const auto outOfLineBytes =
+      data_->stringAllocator().retainedSize() - outOfLineFreeBytes;
+  const int64_t outOfLineBytesPerRow = outOfLineBytes / currentDataRows;
+  const int64_t flatInputBytes = input->estimateFlatSize();
+
+  const auto& spillConfig = spillConfig_.value();
+  // Test-only spill path.
+  if (currentDataRows > 0 && spillConfig.testSpillPct &&
+      (folly::hasher<uint64_t>()(++spillTestCounter_)) % 100 <=
+          spillConfig.testSpillPct) {
+    const int64_t rowsToSpill = std::max<int64_t>(1, currentDataRows / 10);
+    spill(
+        currentDataRows - rowsToSpill,
+        outOfLineBytes - (rowsToSpill * outOfLineBytesPerRow));
+    return;
+  }
+
+  // 1. If current memory usage exceeds spilling threshold, trigger spilling.
+  const auto currentMemoryUsage = memoryPool_->currentBytes();
+  if (spillMemoryThreshold_ != 0 &&
+      currentMemoryUsage > spillMemoryThreshold_) {
+    const int64_t bytesToSpill =
+        currentMemoryUsage * spillConfig.spillableReservationGrowthPct / 100;
+    auto rowsToSpill = std::max<int64_t>(
+        1, bytesToSpill / (data_->fixedRowSize() + outOfLineBytesPerRow));
+    spill(
+        std::max<int64_t>(0, currentDataRows - rowsToSpill),
+        std::max<int64_t>(
+            0, outOfLineBytes - (rowsToSpill * outOfLineBytesPerRow)));
+    return;
+  }
+
+  // 2. If we have enough free rows for input rows and enough variable length
+  // free space for the vector's flat size, no need for spilling.
+  if (freeRows > input->size() &&
+      (outOfLineBytes == 0 || outOfLineFreeBytes >= flatInputBytes)) {
+    return;
+  }
+
+  // For variable length data, we take the flat size of the input as the cap.
+  const int64_t estimatedIncrementalBytes =
```

Comment:
Please use the original variable names in order by to keep the code the same. You can do cleanup and we can review that part separately